drivers/rtc/rtc_rv3028.c

```diff
@@ -0,0 +1,1101 @@
+
+#ifdef CONFIG_RTC_UPDATE
+		if ((status & RV3028_STATUS_UF) != 0 && data->update_callback != NULL) {
+			status &= ~(RV3028_STATUS_UF);
+			update_callback = data->update_callback;
+			update_user_data = data->update_user_data;
+		}
+#endif /* CONFIG_RTC_UPDATE */
+
+		err = rv3028_write_reg8(dev, RV3028_REG_STATUS, status);
+		if (err) {
+			goto unlock;
+		}
+
+		/* Check if interrupt occurred between STATUS read/write */
+		err = rv3028_read_reg8(dev, RV3028_REG_STATUS, &status);
+		if (err) {
+			goto unlock;
+		}
+
+		if (((status & RV3028_STATUS_AF) != 0U && alarm_callback != NULL) ||
+		    ((status & RV3028_STATUS_UF) != 0U && update_callback != NULL)) {
+			/*
+			 * Another interrupt occurred while servicing this one, process current
+			 * callback(s) and yield.
+			 */
+			k_sem_give(&data->int_sem);
+		}
+
+unlock:
+		k_mutex_unlock(&data->lock);
+
+		if (alarm_callback != NULL) {
+			alarm_callback(dev, 0U, alarm_user_data);
+			alarm_callback = NULL;
+		}
+
+		if (update_callback != NULL) {
+			update_callback(dev, update_user_data);
+			update_callback = NULL;
+		}
+	}
+}
+
+static void rv3028_isr_callback_handler(const struct device *port, struct gpio_callback *cb,
+                                        gpio_port_pins_t pins)
+{
+	struct rv3028_data *data = CONTAINER_OF(cb, struct rv3028_data, int_callback);
+
+	ARG_UNUSED(port);
+	ARG_UNUSED(pins);
+
+	k_sem_give(&data->int_sem);
+}
+
+#endif /* RV3028_INT_GPIOS_IN_USE */
+
+static int rv3028_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	struct rv3028_data *data = dev->data;
+	uint8_t date[7];
+	int err;
+
+	if (timeptr == NULL ||
+	    (timeptr->tm_year < RV3028_YEAR_OFFSET) ||
+	    (timeptr->tm_year > (RV3028_YEAR_OFFSET + 99))) {
+		LOG_ERR("Invalid time");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	LOG_DBG("Set time: year = %d, mon = %d, mday = %d, wday = %d, hour = %d, "
+	        "min = %d, sec = %d",
+	        timeptr->tm_year, timeptr->tm_mon, timeptr->tm_mday, timeptr->tm_wday,
+	        timeptr->tm_hour, timeptr->tm_min, timeptr->tm_sec);
+
+	date[0] = bin2bcd(timeptr->tm_sec) & RV3028_SECONDS_MASK;
+	date[1] = bin2bcd(timeptr->tm_min) & RV3028_MINUTES_MASK;
+	date[2] = bin2bcd(timeptr->tm_hour) & RV3028_HOURS_24H_MASK;
+	date[3] = bin2bcd(timeptr->tm_wday) & RV3028_WEEKDAY_MASK;
+	date[4] = bin2bcd(timeptr->tm_mday) & RV3028_DATE_MASK;
+	date[5] = bin2bcd(timeptr->tm_mon + RV3028_MONTH_OFFSET) & RV3028_MONTH_MASK;
+	date[6] = bin2bcd(timeptr->tm_year - RV3028_YEAR_OFFSET) & RV3028_YEAR_MASK;
+
+	err = rv3028_write_regs(dev, RV3028_REG_SECONDS, &date, sizeof(date));
+
+	k_mutex_unlock(&data->lock);
+
+	return err;
+}
+
+static int rv3028_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	uint8_t date[7];
+	int err;
+
+	if (timeptr == NULL) {
+		return -EINVAL;
+	}
+
+	err = rv3028_read_regs(dev, RV3028_REG_SECONDS, date, sizeof(date));
+	if (err) {
+		return err;
+	}
+
```

Comment:
If the time is invalid, this function should return -ENODATA, for the RV3028, the POR (power-on-reset) flag should be able to determine that the time is not valid :) There may be a better way to check but POR flag seems perfect :)

Reply:
Done